contrib/epee/src/net_ssl.cpp

```diff
@@ -519,10 +520,16 @@ bool ssl_options_t::handshake(
 
   boost::system::error_code ec = boost::asio::error::would_block;
   socket.async_handshake(type, boost::lambda::var(ec) = boost::lambda::_1);
-  while (ec == boost::asio::error::would_block)
+  if (io_service.stopped())
   {
     io_service.reset();
```

Comment:
No. We don't need to call `reset` inside the loop.

> when a previous invocation of these functions returned due to the io_context being stopped or running out of work

Reply:
Yup, there is work queued, forgot about that while reviewing.

I still recommend dropping the `stopped()` conditional. I can't recall any example using that style. I doubt there will ever be any performance advantage - the [current non-windows implementation](https://github.com/boostorg/asio/blob/develop/include/boost/asio/detail/impl/scheduler.ipp#L306) is just a lock acquiration and bool set. The [windows implementation](https://github.com/boostorg/asio/blob/develop/include/boost/asio/detail/win_iocp_io_context.hpp#L88) is basically identical. So the `stopped()` call is simply adding an additional lock acquire/release to avoid ocassionally setting a bool.

Reply:
According to asio docs

> Calling restart() while there are unfinished run(), run_one(), run_for(), run_until(), poll() or poll_one() calls results in undefined behaviour

This is applicable to our case due to having another worker threads.

Reply:
So this is the thing that keeps hurting my head about ASIO design BS in our back-and-forth. Maybe its too late for me to be thinking about all this thread stuff ... 

I'm not sure if this design (in our codebase) can work in entirely "legal" ASIO. If those other worker threads "stop" because of ctrl-c (I _think_ thats possible would have to scan again to verify), this loop here can spin because the restart cannot be properly synchronized with those other threads ...?

As in: (1) thread here returns `stopped() == false`, then (2) `stop()` is called exiting all other work threads before the next `poll_one()` in this thread, and then (3) `poll_one()` in this thread never does any work including handling the timer because it always exits immediately after `stop()` was called. And the pending events aren't cleared until destructing the service object.

I don't think its possible without using `reset()` in the loop, which also isn't valid. Although I can never recall `reset()` doing anything unexpected in that situation either.

Reply:
Or wait, if you did the `stopped()` check in the loop too ... ahhh. ASIO how much I love to hate  thee.

Reply:
Anyway, after re-reviewing. The current PR + `if (stopped()) { return ...; }` in the loop is probably good enough for now.

More Info:

* There is a potential race when `stop()` is issued on another thread after the `stopped()` check, and before the timer or socket callback complete. The loop will infinitely exit `poll_one()` immediately and not do any work.
* I can't recall seeing `stopped()` ever used like this. One issue is that it doesn't necessarily indicate when the other threads left the `run()` method. When `stop()` is called, `stopped()` merely indicates that all threads were notified (but they still might be processing work). So the "undefined behavior" case is still plausibly triggered in this PR.

Fixing this entirely requires a bit more massaging. The `if (stopped()) { reset(); }` before the loop doesn't work because its attempting to handle two different cases. In one scenario, theres background worker threads, in the other there are not. The `reset()` call _probably_ has to be done somewhere else externally for the non-background thread case.

Reply:
Updated addressing `a potential race when stop() is issued on another thread` case